src/test/java/build/buildfarm/BuildFarmServerTest.java

```diff
@@ -148,4 +170,130 @@ public void listOperations() {
 
     assertThat(response.getOperationsList()).isEmpty();
   }
+
+  @Test
+  public void canceledOperationIsNoLongerOutstanding() throws RetryException, InterruptedException {
+    Operation operation = executeAction(createSimpleAction());
+
+    // should appear in outstanding list
+    ListOperationsRequest listRequest = ListOperationsRequest.newBuilder()
+        .setName("memory/operations")
+        .setPageSize(1024)
+        .build();
+
+    OperationsGrpc.OperationsBlockingStub operationsStub =
+        OperationsGrpc.newBlockingStub(inProcessChannel);
+
+    ListOperationsResponse listResponse = operationsStub.listOperations(listRequest);
+
+    assertThat(Iterables.transform(listResponse.getOperationsList(), o -> o.getName())).containsExactly(operation.getName());
+
+    CancelOperationRequest cancelRequest = CancelOperationRequest.newBuilder()
+        .setName(operation.getName())
+        .build();
+
+    operationsStub.cancelOperation(cancelRequest);
+
+    // should now be gone
+    listResponse = operationsStub.listOperations(listRequest);
+
+    assertThat(listResponse.getOperationsList()).isEmpty();
+  }
+
+  @Test
+  public void cancellingExecutingOperationFailsPoll()
```

Comment:
nit: it seems you are testing three things for a canceled operation:
- poll should fail
- list should not contain it
- get should contain it with error = CANCELLED
but you do it in two unit tests. Wouldn't just the last one do?

Reply:
duly nitted - moved the operation state check of a getOperation response into its own test with a precondition validation.

At some point I'm going to want to have a full FSM verification for the operation states for each instance type